### PR TITLE
Adds the capability of provisioning with private ip for digitalocean

### PR DESCRIPTION
**Pull request by ufukty**
**Summary**

- Adds the capability of provisioning with private ip for digitalocean.
- Creates an argument which named as `connect_with_private_ip` to make provisioners to use the private ip instead of the public ip.
- Creates another argument which named as `vpc_uuid` for enabling the VPC UUID support that is presented by official DO API 

**Original problem**

Packer is losing connection after provisioner creates a firewall rule that drops the ssh connections coming from outside of local network.

**Done**
- `make generate`
- `make lint` displayed an error:
    ```sh
    $ make lint PKG_NAME=builder/digitalocean
    # Pinning golangci-lint at v1.23.8 as --new-from-rev seems to work properly; the latest 1.24.0 has caused issues with memory consumption
    ==> Updating linter dependencies...
    golangci/golangci-lint info checking GitHub for tag 'v1.23.8'
    golangci/golangci-lint info found version: 1.23.8 for v1.23.8/darwin/amd64
    golangci/golangci-lint info installed /Users/ufuktan/go/bin/golangci-lint
    golangci-lint run ./builder/digitalocean/...
    builder/digitalocean/config.go:183:11: S1028: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (gosimple)
    				errs, errors.New(fmt.Sprintf("user_data_file not found: %s", c.UserDataFile)))
    				      ^
    builder/digitalocean/config.go:194:41: S1028: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (gosimple)
    			errs = packer.MultiErrorAppend(errs, errors.New(fmt.Sprintf("invalid tag: %s", t)))
    			                                     ^
    make: *** [lint] Error 1
    
    ```

### CODE REVIEW

File: builder/digitalocean/config.go [hunk 1]
```diff
@@ -187,6 +195,20 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
 		}
 	}
 
+	// Check if the PrivateNetworking is enabled by user before use VPC UUID
```
Comment (SwampDragons) [inline builder/digitalocean/config.go:198]: I think it makes sense to add a few tests to builder/digitalocean/builder_test.go to make sure this template validation logic works as expected.

**Review by SwampDragons — CHANGES_REQUESTED** [review]
Hi, this looks good! I'd love if you could add a couple of tests to the template logic, but otherwise I think this is good to merge.

**Comment by ufukty** [comment]
Thank you. Tests you want have been added.

**Comment by ufukty** [comment]
circle-ci complaining about whitespace changes at some files that are irrelevant with this pr. I re-cloned the repo and re-ran `make generate-check`. Errors seems gone for me.

https://app.circleci.com/pipelines/github/hashicorp/packer/7625/workflows/503eb522-94ce-4361-a027-8193248aa519/jobs/87592/steps

```
==> Checking that auto-generated code is not changed...
warning: CRLF will be replaced by LF in builder/azure/arm/azure_error_response_test.TestAzureErrorNestedShouldFormat.approved.txt.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in builder/azure/arm/azure_error_response_test.TestAzureErrorSimpleShouldFormat.approved.txt.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in test/fixtures/provisioner-ansible/largish-file.txt.
The file will have its original line endings in your working directory
diff --git a/builder/azure/arm/azure_error_response_test.TestAzureErrorNestedShouldFormat.approved.txt b/builder/azure/arm/azure_error_response_test.TestAzureErrorNestedShouldFormat.approved.txt
index 9160d5bc3..39d4deeec 100644
--- a/builder/azure/arm/azure_error_response_test.TestAzureErrorNestedShouldFormat.approved.txt
+++ b/builder/azure/arm/azure_error_response_test.TestAzureErrorNestedShouldFormat.approved.txt
@@ -1,4 +1,4 @@
-ERROR: -> DeploymentFailed : At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.
-ERROR:   -> BadRequest
-ERROR:   -> InvalidRequestFormat : Cannot parse the request.
-ERROR:     -> InvalidJson : Error converting value "playground" to type 'Microsoft.WindowsAzure.Networking.Nrp.Frontend.Contract.Csm.Public.IpAllocationMethod'. Path 'properties.publicIPAllocationMethod', line 1, position 130.
+ERROR: -> DeploymentFailed : At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.
+ERROR:   -> BadRequest
+ERROR:   -> InvalidRequestFormat : Cannot parse the request.
+ERROR:     -> InvalidJson : Error converting value "playground" to type 'Microsoft.WindowsAzure.Networking.Nrp.Frontend.Contract.Csm.Public.IpAllocationMethod'. Path 'properties.publicIPAllocationMethod', line 1, position 130.
diff --git a/builder/azure/arm/azure_error_response_test.TestAzureErrorSimpleShouldFormat.approved.txt b/builder/azure/arm/azure_error_response_test.TestAzureErrorSimpleShouldFormat.approved.txt
index d50bdc052..4b4834c62 100644
--- a/builder/azure/arm/azure_error_response_test.TestAzureErrorSimpleShouldFormat.approved.txt
+++ b/builder/azure/arm/azure_error_response_test.TestAzureErrorSimpleShouldFormat.approved.txt
@@ -1 +1 @@
-ERROR: -> ResourceNotFound : The Resource 'Microsoft.Compute/images/PackerUbuntuImage' under resource group 'packer-test00' was not found.
+ERROR: -> ResourceNotFound : The Resource 'Microsoft.Compute/images/PackerUbuntuImage' under resource group 'packer-test00' was not found.
diff --git a/test/fixtures/provisioner-ansible/largish-file.txt b/test/fixtures/provisioner-ansible/largish-file.txt
index 4df7b0668..0dec1344a 100644
Binary files a/test/fixtures/provisioner-ansible/largish-file.txt and b/test/fixtures/provisioner-ansible/largish-file.txt differ
Found diffs in go generated code.
You can use the command: `make generate` to reformat code.
make: *** [Makefile:132: generate-check] Error 1
```

**Comment by SwampDragons** [comment]
I just reran the test suite and you're right it was some unrelated flake. Thanks for this contribution!

**Pull request merged**